open_flamingo/train/train.py

```diff
@@ -252,8 +268,54 @@ def main():
     if args.offline:
         os.environ["WANDB_MODE"] = "offline"
         os.environ["TRANSFORMERS_OFFLINE"] = "1"
+
     args.local_rank, args.rank, args.world_size = world_info_from_env()
-    device_id = init_distributed_device(args)
+    if args.deepspeed:
+        torch.cuda.set_device(args.local_rank)
+        device = torch.device("cuda", args.local_rank)
+        # Initializes the distributed backend which will take care of sychronizing nodes/GPUs
+        deepspeed.init_distributed()
+
+        zero_opt_dict = {
+            "stage": args.deepspeed_stage,
+            "overlap_comm": True,
+            "contiguous_gradients": True,
+            "offload_param": {"device": "none"},  # TODO: Support CPU offload
+            "offload_optimizer": {"device": "none"},
+            "stage3_param_persistence_threshold": 1e4,
+            "stage3_max_live_parameters": 3e7,
+            "stage3_prefetch_bucket_size": 3e7,
+            "stage3_gather_16bit_weights_on_model_save": True,
```

Comment:
Are all model weights saved in 16bit, regardless of the precision flag?

Reply:
This saves additional 16 bit weights is for stage3 but is unnecessary and will be slow! There is a script to 'reconstruct' the fp32 weights from the checkpoint as described [here](https://huggingface.co/docs/accelerate/usage_guides/deepspeed#saving-and-loading).

Reply:
I see -- so since we save ckpts fairly often, should we should set this flag to False by default,  and then provide a separate script to all_gather the weights into one file offline? (I guess wandb ckpt saving will need to be turned off)

Reply:
Deepspeed auto creates the script in the checkpoint dir which is nice :). Good point on wandb. For stage 3 yes although we can also just save the sharded checkpoint.